## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [2]:
import pandas as pd

df = pd.read_csv('/Users/mac/Downloads/Starter_Code 2/Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [3]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [4]:
# Use the describe function to gather some basic statistics
statistics = df.describe()
statistics


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data
import numpy as np



df.rename(columns={'job': 'occupation'})
df



first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  
0                1  
1                0  
2                6  
3                3  
4                1  
...            ...  
54634            8  
54635            0  
54636            6  
54637            0  
54638            5  

[54639 rows x 18 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [7]:
df_organized = df[['last', 'client_id','job', 'category', 'subcategory', 'unit_price', 'unit_cost', 'qty']]
df_organized

last  client_id                      job     category  \
0        Harding      58515             Immunologist        decor   
1          Myers      37609          Music therapist  consumables   
2         Watson      57113             Immunologist     software   
3          Baker      46554    Accounting technician  consumables   
4       Schwartz      92089             Chiropractor  consumables   
...          ...        ...                      ...          ...   
54634      Reese      29043           Radio producer     software   
54635      Moore      41908                 Musician  consumables   
54636  Gutierrez      35176         Graphic designer  consumables   
54637    Solomon      24485             Toxicologist    furniture   
54638       Bell      24007  Chief Financial Officer     software   

              subcategory  unit_price  unit_cost   qty  
0                wall art     1096.80     762.71   105  
1                    pens       24.95      15.09    21  
2      project management       13.52       7.86    39  
3                    pens       36.42      24.85    29  
4                    misc      195.10     108.17    20  
...                   ...         ...        ...   ...  
54634           logistics       83.13      51.60    33  
54635    kitchen supplies      206.59     175.46    47  
54636   bathroom supplies       65.66      57.31   475  
54637              stands        1.48       1.22   112  
54638  project management        3.01       1.61  1031  

[54639 rows x 8 columns]

In [8]:
# What three item categories had the most entries?
df_amount = df_organized['category'].value_counts(ascending=False).head(3)
print(df_amount)

#consumables, furniture, software.


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [9]:
# For the category with the most entries, which subcategory had the most entries?
# bathroom supplies did.
#df_organized[['category', 'subcategory']]

df_amount = df_organized.loc[df_organized['category']=='consumables']['subcategory'].value_counts(ascending=False).head(1)
df_amount



subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [10]:
# Which five clients had the most entries in the data?
df_client_amount = df_organized['client_id'].value_counts(ascending=False).head(5)
df_client_amount


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [11]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = df_client_amount.index.to_list()
top_5_clients


[33615, 66037, 46820, 38378, 24741]

In [12]:
sorted = df_organized.sort_values(by= ['qty'], ascending=False)
sorted

last  client_id                                  job     category  \
21801  Anderson      78965                  Exhibition designer    furniture   
25972   Stewart      86721            Psychologist, educational    furniture   
8761     Watson      30143                    Ceramics designer  consumables   
15112   Hawkins      85931  Social research officer, government  consumables   
50048   Burnett      77354                      Records manager  consumables   
...         ...        ...                                  ...          ...   
26466    Golden      18452                        Buyer, retail  consumables   
51020    Potter      86040                          Adult nurse     software   
11742  Mcdaniel      75225        Restaurant manager, fast food  consumables   
8267      Reyes      28523     Production assistant, television    furniture   
21090    Bowman      72892                        Archaeologist  electronics   

            subcategory  unit_price  unit_cost      qty  
21801            stands      147.66     100.11  3958244  
25972            tables      100.40      73.05   977121  
8761               pens       24.95      15.09   842581  
15112    paper products      323.42     183.84   637318  
50048  kitchen supplies       81.23      69.22   494692  
...                 ...         ...        ...      ...  
26466  kitchen supplies      251.68     141.70        0  
51020        accounting        2.67       2.35        0  
11742  kitchen supplies       37.14      27.50        0  
8267            seating       34.02      20.43        0  
21090           cabling       96.86      74.98        0  

[54639 rows x 8 columns]

In [13]:
# How many total units (the qty column) did the client with the most entries order order?
sorted_tot = sorted.loc[sorted['client_id'] == 33615]['qty'].sum()
sorted_tot

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [14]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty

df['line_subtotal'] = df['unit_price']* df['qty']
df[['line_subtotal', 'unit_price', 'qty']].head(2)


line_subtotal  unit_price  qty
0      115164.00     1096.80  105
1         523.95       24.95   21

In [15]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

df['total_weight'] = df['unit_weight'] * df['qty']


def calculate_ship_price(total_weight):
    if total_weight > 50:
        return 7 * total_weight
    else: 
        return total_weight * 10
df["ship price"] = df['total_weight'].apply(calculate_ship_price)
df[['unit_price', 'unit_weight', 'qty', 'total_weight','ship price']].head(3)
    


unit_price  unit_weight  qty  total_weight  ship price
0     1096.80         7.50  105        787.50     5512.50
1       24.95         1.49   21         31.29      312.90
2       13.52         1.68   39         65.52      458.64

In [16]:
# line price = line  total _ship rice
df['line_price']= (df['line_subtotal'] + df['ship price'])* 1.0925
df[['line_subtotal', 'ship price', 'line_price']].head(3)


line_subtotal  ship price     line_price
0      115164.00     5512.50  131839.076250
1         523.95      312.90     914.258625
2         527.28      458.64    1077.117600

In [17]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['ship price']
df.head(3)



first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  ship price     line_price  \
0            1      115164.00        787.50     5512.50  131839.076250   
1            0         523.95         31.29      312.90     914.258625   
2            6         527.28         65.52      458.64    1077.117600   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [18]:
# Create a column for the profit of each line using line cost and line price

df["line_profit"] = df['line_price'] -df['line_cost']
df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  ship price     line_price  line_cost  \
0      115164.00        787.50     5512.50  131839.076250   85597.05   
1         523.95         31.29      312.90     914.258625     629.79   
2         527.28         65.52      458.64    1077.117600     765.18   

    line_profit  
0  46242.026250  
1    284.468625  
2    311.937600  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [19]:
total_sum = df.loc[df['order_id'].isin([2742071, 2173913, 6128929]), 'line_price'].sum()
total_sum 

1238641.8568250001

In [20]:
# Check your work using the totals above
order1 = round(df.loc[df['order_id']==2742071,'line_price'].sum(),2)
order2 = round(df.loc[df['order_id']==2173913, 'line_price'].sum(),2)
order3 = round(df.loc[df['order_id']==6128929, 'line_price'].sum(),2)
print(f'Order 2742071 total:${order1}')
print(f'Order 2173913 total:${order2}')
print(f'Order 6128929 total:${order3}')

Order 2742071 total:$152811.9
Order 2173913 total:$162388.72
Order 6128929 total:$923441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [21]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

for client in top_5_clients:
    top_5_spent = round(df.loc[df['client_id']== client ,'line_price'].sum(),2)
    print(f'{client} : ${top_5_spent}')

33615 : $8377308.57
66037 : $10259514.8
46820 : $9743794.32
38378 : $12906550.88
24741 : $82268891.98


In [49]:


top_5_clients_totals = df[df['client_id'].isin(top_5_clients)].groupby('client_id').agg({
    'qty': 'sum',
    'ship price': 'sum',
    'line_price': 'sum',
    'line_cost': 'sum',
    'line_profit': 'sum'
}).reset_index()

top_5_clients_totals[['client_id', 'qty', 'ship price', 'line_price', 'line_cost', 'line_profit']] 


client_id     qty  ship price    line_price    line_cost   line_profit
0      24741  239862  5126448.37  8.226889e+07  45688899.71  3.657999e+07
1      33615   64313  1828984.89  8.377309e+06   6175313.91  2.201995e+06
2      38378   73667  3429455.40  1.290655e+07   9634720.98  3.271830e+06
3      46820   75768  1601448.84  9.743794e+06   7007191.64  2.736603e+06
4      66037   43018  1395151.85  1.025951e+07   7004482.98  3.255032e+06

In [50]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.
summary_df = pd.DataFrame({
    'client_id': top_5_clients_totals['client_id'],
    'Units ': top_5_clients_totals['qty'],
    'total shipping price (millions)': top_5_clients_totals['ship price'],
    'total revenue (millions)': top_5_clients_totals['line_price'],
    'total cost (millions)': top_5_clients_totals['line_cost'],
    'total profit (millions)': top_5_clients_totals['line_profit']
})
summary_df

client_id  Units   total shipping price (millions)  \
0      24741  239862                       5126448.37   
1      33615   64313                       1828984.89   
2      38378   73667                       3429455.40   
3      46820   75768                       1601448.84   
4      66037   43018                       1395151.85   

   total revenue (millions)  total cost (millions)  total profit (millions)  
0              8.226889e+07            45688899.71             3.657999e+07  
1              8.377309e+06             6175313.91             2.201995e+06  
2              1.290655e+07             9634720.98             3.271830e+06  
3              9.743794e+06             7007191.64             2.736603e+06  
4              1.025951e+07             7004482.98             3.255032e+06

In [51]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_colums = df['ship price', 'line_price', 'line_cost', 'line_profit']


# Define a function that converts a dollar amount to millions.
def currency_format_millions(df, money_colums):
    if df[money_columscolumns]/= 1_000_000

    return f'${x:.2f}M'


# Apply the currency_format_millions function to only the money columns. 
summary_df[money_columns] = summary_df[money_columns].applymap(currency_format_millions)


# Rename the columns to reflect the change in the money format. 
summary_df = pd.DataFrame({
    'client_id': top_5_clients_totals['client_id'],
    'Units ': top_5_clients_totals['qty'],
    'total shipping price (millions)': top_5_clients_totals['ship price'],
    'total revenue (millions)': top_5_clients_totals['line_price'],
    'total cost (millions)': top_5_clients_totals['line_cost'],
    'total profit (millions)': top_5_clients_totals['line_profit']
})
summary_df





KeyError: ('ship price', 'line_price', 'line_cost', 'line_profit')

In [ ]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
